Feature autosummary packages

woutdenolf opened the pull request:
Subject: recursively add modules and sub-packages to a package with autosummary

### Feature or Bugfix
<!-- please choose -->
- Feature

### Purpose
- Add two new items to the module template: "modules" and "packages".
- This allows to automatically add modules and sub-packages for modules that are packages.
- It works recursively.

For example a [project](https://github.com/woutdenolf/sphinx_example):
```script
package/
|-- __init__.py
|-- foo.py
|
|-- subpackage/
|   |-- __init__.py
|   |-- foo.py
|
|-- doc/
|   |-- source
|       |--index.rst
|       |--conf.py
```
with this index.rst:
```script
API Reference
=============

.. rubric:: Packages

.. autosummary::
   :toctree: apidocs

    package

```
will generate
```script
doc/source/apidocs
|-- package.rst
|
|-- modules/
|   |-- package.foo.rst
|
|-- packages/
|   |-- package.subpackage.rst
|   |-- modules/
|       |-- package.subpackage.foo.rst
```
which renders as

![2019-02-08_17-22-55](https://user-images.githubusercontent.com/7264703/52491496-8ae10600-2bc7-11e9-94c4-1a219505d2af.png)


### Detail
- Recursion is disabled by default (old behavior)
- Parameter to control recursion depth
- Exclude/include pattern is not supported but may be useful

### Relates
- issue #248

woutdenolf committed:
[autosummary] add modules and packages for modules

woutdenolf committed:
[autosummary] tests for autosummary modules and packages

woutdenolf committed:
[autosummary] import pkgutil before sys

woutdenolf committed:
[autosummary] import pkgutil first

woutdenolf committed:
[autosummary] fix some flake8 issues

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -236,6 +236,19 @@ The following variables available in the templates:
+   List containing names of "public" modules in the package.  Only available for
+   modules that are packages.
+
+   .. versionadded:: 2.0.0
```
We usually use `2.0` instead for versionadded.

woutdenolf replied:
Done

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -188,6 +189,29 @@ def get_members(obj, typ, include_public=[], imported=True):
                           if x in include_public or not x.startswith('_')]
                 return public, items
 
+            def get_package_members(out_dict, include_public=[], imported=True):
```
Why did you not use return value? I feel `out_dict` is strange interface.

In addition, is `include_public` needed? It seems not given.

tk0miya replied:
Please add type annotations here.

woutdenolf replied:
Type annotation added. `get_members` had an unused `include_public` so I added it to `get_package_members`  as well. If we remove it we should do it for both.

tk0miya replied:
No, `include_public` for `get_members() has been used. So we can't remove it.
https://github.com/sphinx-doc/sphinx/blob/4628264551a2c2ef16213abdd6965c7baf65c536/sphinx/ext/autosummary/generate.py#L205-L206

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -188,6 +189,29 @@ def get_members(obj, typ, include_public=[], imported=True):
+                        warn('[autosummary] failed to import %s: %s' % (fullname, e))
+                        continue
+                    if ispkg:
+                        kpublic, kall = 'packages', 'all_packages'
```
What does "k" prefix mean?

woutdenolf replied:
Refactored

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -188,6 +189,29 @@ def get_members(obj, typ, include_public=[], imported=True):
+                if not hasattr(obj, '__path__'):
+                    return
+                out_dict['modules'], out_dict['all_modules'] = [], []
+                out_dict['packages'], out_dict['all_packages'] = [], []
```
When are `all_modules` and `all_packages` used?

woutdenolf replied:
Analog to `all_functions`, `all_classes`, `all_methods`, etc... I suppose they are only used when you specify them in your template.

tk0miya replied:
Reasonable.

tk0miya reviewed (changes requested):
Great work! LGTM with nits.

Note: we have to check whole of document for autosummary again.

tk0miya commented:
I noticed autosummary does not detect new submodules after a build once. Because it skip to process if .rst files are already generated. To create right document, we need to clear generated .rst files before building.

woutdenolf committed:
[autosummary] do not skip an existing source when it is a package

woutdenolf committed:
[autosummary] add recursion limit config parameter

woutdenolf committed:
[main] allow for docstring extraction

woutdenolf committed:
[autosummary] modify docs

woutdenolf committed:
[autosummary] test package recursion

woutdenolf committed:
Merge branch 'master' into feature-autosummary-packages

woutdenolf commented:
Proposal to deal with new and removed submodules and subpackages: in `generate_autosummary_docs` existing .rst files are overwritten when they correspond to a package.

Because this is not always what you want (you may have a user created .rst for a package already), I also propose to add the config parameter which enables/disables package recursion. Instead of a boolean I used an integer so you can additionally control the recursion depth:

- `autosummary_recursion_limit == 0`: disable recursion (default, old behavior)
- `autosummary_recursion_limit < 0`: unlimited recursion
- `autosummary_recursion_limit > 0` limited recursion depth

Caveat: generated .rst files are not deleted when rebuilding with a decreased recursion limit, which leads to warnings about rst files not being included in the toctree.

Alternatively one could somehow mark the .rst files generated by autosummary to distinguish them from user generated files, so you would know whether they can be deleted (in case of decreased recursion limit) and overwritten (in case of package recursion).

woutdenolf committed:
[autosummary] small flake8/mypy bugs

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -25,14 +25,19 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
    These files by default contain only the corresponding
    :mod:`sphinx.ext.autodoc` directive, but can be customized with templates.
 
+3. Optionally, the new :confval:`autosummary_recursion_limit` config value can be
```
The term "new" is not needed here.

tk0miya replied:
I fixed "new" in the list item above: https://github.com/sphinx-doc/sphinx/pull/6104

woutdenolf replied:
done

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -25,14 +25,19 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
 
+3. Optionally, the new :confval:`autosummary_recursion_limit` config value can be
+   used to generate "stub" files recursively for packages. Warning: this will overwrite
+   existing files for packages.
```
`autosummary_generate` option always overwrite files. not only files for packages. Is this warning needed?

woutdenolf replied:
removed the warning

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -54,6 +59,27 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
 
        .. currentmodule:: sphinx.ext.autosummary
 
+   When :confval:`autosummary_recursion_limit != 0`, the :rst:dir:`autosummary`
```
The `confval` role takes the variable name as a content. And `!= 0` is not a variable name. So this marking up is not good. How about ```When :confval:`autosummary_recursion_limit` is not 0, ...``` instead?

woutdenolf replied:
paragraph removed altogether

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -54,6 +59,27 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
+
+   For example, ::
+
+       .. currentmodule:: sphinx
```
`sphinx.ext` ?

woutdenolf replied:
paragraph removed altogether

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -136,13 +162,25 @@ also use this new config value:
    The new files will be placed in the directories specified in the
    ``:toctree:`` options of the directives.
 
+Packages can be explored recursively when generating stub pages.
```
I think this is a part of description for `autosummary_recursion_limit`. So this should be placed under the entry.

woutdenolf replied:
removed

tk0miya commented on sphinx/__main__.py:
```diff
@@ -12,4 +12,5 @@
 
-sys.exit(main(sys.argv[1:]))
+if __name__ == "__main__":
+    sys.exit(main(sys.argv[1:]))
```
Is this change related with this PR?

tk0miya commented on sphinx/ext/autosummary/__init__.py:
```diff
@@ -731,7 +734,7 @@ def setup(app):
+    app.add_config_value('autosummary_recursion_limit', 0, True, [int])
     app.add_config_value('autosummary_mock_imports',
                          lambda config: config.autodoc_mock_imports, 'env')
-
```
why removing this line?

woutdenolf replied:
sorry, put it back

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -90,12 +91,19 @@ def _underline(title, line='='):
+                              builder=None,            # type: Builder
+                              template_dir=None,       # type: str
+                              imported_members=False,  # type: bool
+                              recursion_limit=0,       # type: int
```
Could you add this to the tail of argument list to keep compatibility? I'm afraid of changing the order of parameters.

woutdenolf replied:
done

tk0miya commented on tests/roots/test-ext-autosummary-package/conf.py:
```diff
@@ -0,0 +1,10 @@
+autosummary_generate = True
+
+# The suffix of source filenames.
+source_suffix = '.rst'
```
Since 2.0, the default setting of `source_suffix` becomes `.rst`. So this setting is not needed.

woutdenolf replied:
done

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -151,11 +159,32 @@ def generate_autosummary_docs(sources, output_dir=None, suffix='.rst',
+        ispackage = hasattr(obj, '__path__')
+        if ispackage:
+            depth = len(name.split('.')) - 1
+            recurse_package = depth < recursion_limit or recursion_limit < 0
```
It seems `recursion_limit` is compared with the depth of package from root. I think it is not called "recursion".

woutdenolf replied:
Yes I should specify in the docs that this is recursion depth from root. At first I was thinking of limiting the `generate_autosummary_docs` call depth but I think this gives confusing results, depending on the situation. Let me know if you have an issue with the concept or just the name of the parameter. Not sure what name would make more sense `autosummary_depth_limit`, `autosummary_package_depth_limit`, ...

woutdenolf replied:
Called it `autosummary_depth_limit`. Docs mention it is the depth starting from root.

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -54,6 +59,27 @@ The :mod:`sphinx.ext.autosummary` extension does this in two parts:
+   When :confval:`autosummary_recursion_limit != 0`, the :rst:dir:`autosummary`
+   directive recursively generates stub ``.rst`` files for packages.
+
+   For example, ::
```
Is this example related with above sentence? It seems `autosummary_recursion_limit` is not set on this document.

woutdenolf replied:
paragraph removed altogether

tk0miya reviewed (commented):
I reviewed partially (I'll review again later if i have time)

woutdenolf committed:
Merge branch 'master' into feature-autosummary-packages

woutdenolf committed:
[autosummary] address code review

woutdenolf committed:
Merge branch 'master' into feature-autosummary-packages

shimizukawa commented:
My opinion is:

* +1 for recursive.
* -1 for limitation for recursive. I don't 
* -1 for separated templates of module and package. I think users don't care whether namespace is module or package.

woutdenolf commented:
Thanks for the feedback!

> * -1 for limitation for recursive.

I agree limiting recursion depth may not be that useful. I would personally use it to eliminate the test suites from my API docs, but that's just because all of my test suites are in subpackages at the same depth, while no other subpackage exists at this depth. Exclusion patterns may be a better way to do this though. Maybe for another PR if this one gets accepted.

> * -1 for separated templates of module and package. I think users don't care whether namespace is module or package.

I find module/package separation cleaner, but that's just personal taste.

If we can get a second opinion on these two things (@tk0miya?) I will make changes accordingly.

woutdenolf committed:
Merge branch 'master' into feature-autosummary-packages

woutdenolf committed:
[autosummary] doclint violation

tk0miya commented:
>-1 for limitation for recursive.

I don't have opinion for this. But I only feel depth limitation *from specified level* might be useful (ex. limitation 2 levels from `sphinx.environment`), not from root. It would be better if exclusion is added instead.

>-1 for separated templates of module and package. I think users don't care whether namespace is module or package.

I don't understand which is better. From author side, modules and packages are different. But it would not be important for readers. IMO, the difference is not important for users of the library. For example, I misunderstood `os` is a package. But the difference does not harm me ever.
Is there any advantage to separate them?

woutdenolf committed:
[autosummary] remove recursion limit and module/package separation

woutdenolf committed:
[autosummary] forgot to adapt types

woutdenolf commented:
I removed the recursion limit (`autosummary_recursive` is now a boolean) and removed the separation between modules and packages (all end up under the `modules` rubric).

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -168,11 +176,26 @@ def generate_autosummary_docs(sources, output_dir=None, suffix='.rst',
+                # Warning: this file could have been created by the user
+                # in which case it is lost.
+                warn('[autosummary] overwriting docs of package %s' % (repr(name)))
+                os.remove(fn)
```
I will add an option for overwriting stub files in #6422.
So this special handling is not needed.

woutdenolf replied:
Merged with master.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -201,6 +224,21 @@ def get_members(obj, types, include_public=[], imported=True):
                           if x in include_public or not x.startswith('_')]
                 return public, items
 
+            def get_modules(obj, include_public=[]):
```
It seems `include_public` is never given. Why is this needed?

woutdenolf replied:
Same goes for the already existing `get_members`, I just stayed consistent.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -201,6 +224,21 @@ def get_members(obj, types, include_public=[], imported=True):
+                for _, modname, ispkg in pkgutil.iter_modules(obj.__path__):
+                    fullname = name + '.' + modname
+                    try:
+                        import_by_name(fullname)
```
Why do you check the module is importable here? No check is needed here to me.

woutdenolf replied:
Removed it

tk0miya commented on sphinx/ext/autosummary/templates/autosummary/module.rst:
```diff
@@ -34,3 +34,15 @@
+.. rubric:: modules
+
+.. autosummary::
+   :toctree: modules
```
Is output directory fixed?

woutdenolf replied:
What else do you suggest? Some options one can think of:
 - specify nothing `:toctree:`: all files are generated in the same directory
 - use package name `:toctree: {{ name }}`: might be confusing (dir tree does not reflect package/subpackage structure)
 - inherit parent toctree `:toctree: {{ ??? }}`: Needs another context variable added by  `generate_autosummary_content`. Not sure how to retrieve the parent toctree name.

tk0miya replied:
At I commented first, I thought inheriting parent toctree is better. But, now, I feel it seems no reasons to make a sub directory. Why `generated/modules/modules/modules/...` directories are needed? IMO, same directory is better to me.

tk0miya commented on tests/test_ext_autosummary.py:
```diff
@@ -222,6 +224,115 @@ def test_autosummary_generate(app, status, warning):
+
+
+@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive')
+def test_autosummary_recursive(make_app, app_params):
```
I feel this testcase is too complicated... How about making it simple?

I think main concern of this testcase is recursive generation. So it is fine to just check existence of stub files.
And it would be great if check root stub file contains `autosummary` definition for recursive call.
```
@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive', confoverrides={'autosummary_recursive': True})
def test_autosummary_recursive_enabled(...):
    app.build()
    assert (app.srcdir / 'generated' / 'package.rst').exists()
    assert (app.srcdir / 'generated' / 'package.module.rst').exists()
    ...

    content = (app.srcdir / 'generated' / 'package.rst').text()
    assert '.. autosummary::\n  :toctree: module\n...' in content


@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive', confoverrides={'autosummary_recursive': True})
def test_autosummary_recursive_disabled(...):
    app.build()
    assert (app.srcdir / 'generated' / 'package.rst').exists()
    assert not (app.srcdir / 'generated' / 'package.module.rst').exists()
    ...

```

woutdenolf replied:
Done

tk0miya commented on tests/roots/test-ext-autosummary-recursive/package/package/package/package/module_importfail.py:
```diff
@@ -0,0 +1,4 @@
+import sys
```
Why is such deeper module needed? I think it is enough to test sub-package.

woutdenolf replied:
Done

tk0miya commented on doc/usage/extensions/autosummary.rst:
```diff
@@ -136,13 +141,20 @@ also use these config values:
    The new files will be placed in the directories specified in the
    ``:toctree:`` options of the directives.
 
+.. confval:: autosummary_recursive
```
It would be also good to add `:recursive:` option to `autosummary` directive. It allows to switch recursive-ness for each autosummary calls. What do you think?

shimizukawa replied:
I don't know what to do without discussing based on the use case.

* What kind of cases do you want to control recursion with common settings?
* What kind of cases do you want to control recursion separately for the autosummary directive?

woutdenolf replied:
@tk0miya Would this require adding something like `sphinx/ext/autodoc/directive.py` to the autosummary extension? I'm not familiar with this aspect but I could try to figure it out.

@shimizukawa This would allow the user to exclude specific packages from being explored when `autosummary_recursive == True`. To work it requires `autosummary_overwrite == False` and the user adding for example `package.excludepackage.rst`. The `excludepackage` package itself would be in the docs though, so I'm not sure what the use case is for this.

tk0miya replied:
As a trial, I implemented :recursive: option. Please see https://github.com/tk0miya/sphinx/tree/6040_autosummary_recursive.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -126,12 +127,19 @@ def render(self, template_name, context):
+                              builder=None,            # type: Builder
+                              template_dir=None,       # type: str
+                              imported_members=False,  # type: bool
+                              recursive=False,         # type: bool
```
Reordering arguments is always breaking change. Could you add a new option to the tail please? I'm afraid that 3rd party application might give parameter as positional.

woutdenolf replied:
Done

woutdenolf committed:
Merge with master and address code review

woutdenolf committed:
remove redundant import check

woutdenolf committed:
refactor autosummary recursion tests

woutdenolf commented:
Lets celebrate the one year anniversary of the PR ;-). Anything I can do? Or should I give up on this?

tk0miya commented on tests/test_ext_autosummary.py:
```diff
@@ -254,6 +255,64 @@ def test_autosummary_generate_overwrite2(app_params, make_app):
+    generated /= toctree
+    assert (generated / 'package.package.package.module.rst').exists()
+    assert not (generated / 'package.package.package.package.rst').exists()
+    if toctree:
```
Why this if condition needed? It seems always true.

tk0miya commented on tests/test_ext_autosummary.py:
```diff
@@ -254,6 +255,64 @@ def test_autosummary_generate_overwrite2(app_params, make_app):
+    assert (generated / 'package.package.package.rst').exists()
+    content = (generated / 'package.package.package.rst').text()
+    assert 'package.package.package.module' in content
+    assert 'package.package.package.package' not in content
```
These deeper packages might be needed for checking the behavior of recursive_limit. But they're no longer needed now.

tk0miya commented on sphinx/ext/autosummary/generate.py:
```diff
@@ -151,6 +153,16 @@ def get_members(obj: Any, types: Set[str], include_public: List[str] = [],
                   if x in include_public or not x.startswith('_')]
         return public, items
 
+    def get_modules(obj: Any, include_public: List[str] = []) -> Tuple[List[str], List[str]]:
```
`include_public` is not used. It should be removed. Same option in `get_members()` is only used for `__init__()` method. No reason to add such option to here.

tk0miya reviewed (changes requested):
Sorry for very very very late response. Let's step forward again for 3.0 release. I believe this is one of key feature of the release!

tk0miya commented:
Now I'm merging this (see #7527). Thank you for your work!
